pallets/parachain-system/src/tests.rs

```diff
@@ -137,15 +137,23 @@ fn send_message(dest: ParaId, message: Vec<u8>) {
 impl XcmpMessageSource for FromThreadLocal {
 	fn take_outbound_messages(maximum_channels: usize) -> Vec<(ParaId, Vec<u8>)> {
 		let mut ids = std::collections::BTreeSet::<ParaId>::new();
-		let mut taken = 0;
+		let mut taken_messages = 0;
+		let mut taken_bytes = 0;
 		let mut result = Vec::new();
 		SENT_MESSAGES.with(|ms| {
 			ms.borrow_mut().retain(|m| {
 				let status = <Pallet<Test> as GetChannelInfo>::get_channel_status(m.0);
-				let ready = matches!(status, ChannelStatus::Ready(..));
-				if ready && !ids.contains(&m.0) && taken < maximum_channels {
+				let ChannelStatus::Ready(max_size_now, max_size_ever) = status else { return false };
```

Comment:
This change is wrong. As this is within a `retain`, returning false (e.g. when the channel is full) would prune the message forever.

Reply:
yes, my bad, pushed a fix

Reply:
I pushed a more comprehensive version + added a bunch more to the test you wrote, just while I was playing around with it.